_webp.c

```diff
@@ -8,8 +8,491 @@
+
+    if (!PyArg_ParseTuple(args, "S", &webp_string)) {
+        return NULL;
+    }
+    PyBytes_AsStringAndSize((PyObject *) webp_string, (char**)&webp, &size);
+    WebPData webp_src = {webp, size};
+
+    // Sniff the mode, since the decoder API doesn't tell us
+    WebPDecoderConfig config;
+    char* mode = "RGBA";
+    if (WebPGetFeatures(webp, size, &config.input) == VP8_STATUS_OK) {
+        if (!config.input.has_alpha) {
+            mode = "RGB";
+        }
+    }
+
+    // Create the decoder (default mode is RGBA, if no options passed)
+    WebPAnimDecoderObject* decp;
+    decp = PyObject_New(WebPAnimDecoderObject, &WebPAnimDecoder_Type);
+    if (decp) {
+        decp->mode = mode;
+        if (WebPDataCopy(&webp_src, &(decp->data))) {
+            WebPAnimDecoder* dec = WebPAnimDecoderNew(&(decp->data), NULL);
+            if (dec) {
+                if (WebPAnimDecoderGetInfo(dec, &(decp->info))) {
+                    decp->dec = dec;
+                    return (PyObject*) decp;
+                }
+            }
+        }
+        PyObject_Del(decp);
+    }
+    fprintf(stderr, "Error! Could not create decoder object.\n");
+    Py_RETURN_NONE;
+}
+
+PyObject* _anim_decoder_dealloc(PyObject* self)
+{
+    WebPAnimDecoderObject* decp = (WebPAnimDecoderObject *)self;
+    WebPDataClear(&(decp->data));
+    WebPAnimDecoderDelete(decp->dec);
+    Py_RETURN_NONE;
+}
+
+PyObject* _anim_decoder_get_info(PyObject* self, PyObject* args)
+{
+    WebPAnimDecoderObject* decp = (WebPAnimDecoderObject *)self;
+    WebPAnimInfo* info = &(decp->info);
+    return Py_BuildValue("IIIIIs",
+        info->canvas_width, info->canvas_height,
+        info->loop_count,
+        info->bgcolor,
+        info->frame_count,
+        decp->mode
+    );
+}
+
+PyObject* _anim_decoder_get_chunk(PyObject* self, PyObject* args)
+{
+    char *mode;
+    PyObject *ret;
+    WebPChunkIterator iter;
+
+    if (!PyArg_ParseTuple(args, "s", &mode)) {
+        return NULL;
+    }
+
+    WebPAnimDecoderObject* decp = (WebPAnimDecoderObject *)self;
+    const WebPDemuxer* demux = WebPAnimDecoderGetDemuxer(decp->dec);
+    if (!WebPDemuxGetChunk(demux, mode, 1, &iter)) {
+        Py_RETURN_NONE;
+    }
+
+    ret = PyBytes_FromStringAndSize((const char*)iter.chunk.bytes, iter.chunk.size);
+    WebPDemuxReleaseChunkIterator(&iter);
+
+    return ret;
+}
+
+PyObject* _anim_decoder_get_next(PyObject* self, PyObject* args)
+{
+    uint8_t* buf;
+    int timestamp;
+    PyObject *bytes;
+
+    WebPAnimDecoderObject* decp = (WebPAnimDecoderObject *)self;
+    if (!WebPAnimDecoderGetNext(decp->dec, &buf, &timestamp)) {
+        fprintf(stderr, "Error! Failed to read next frame.\n");
+        Py_RETURN_NONE;
+    }
+
+    // HACK: If original mode was RGB, we need to strip alpha before passing back, this
+    // is needed because internally WebPAnimDecoder doesn't suppor ta non-alpha mode
+    uint32_t size = decp->info.canvas_width * 4 * decp->info.canvas_height;
```

Comment:
This should probably be a py_ssize_t, and there should be a check for overflow, or some sort of guarantee that this will never overflow due to limits on info.canvas_width.

Reply:
Actually, looking at this, I'm not sure why we're not just returning this as RGBA or RGBX